File: src/test/java/ru/org/linux/user/UserEventDaoIntegrationTest.java
/*
 * Copyright 1998-2015 Linux.org.ru
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package ru.org.linux.user;

import com.google.common.collect.ImmutableList;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.dao.DuplicateKeyException;
import org.springframework.test.context.ContextConfiguration;
import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

import static org.junit.Assert.assertEquals;

@RunWith(SpringJUnit4ClassRunner.class)
@ContextConfiguration(classes = UserEventDaoIntegrationTestConfiguration.class)
@Transactional
public class UserEventDaoIntegrationTest {
  private static final int TEST_TOPIC_ID = 98075;
  private static final int TEST_USER_ID = 32670;

  @Autowired
  private UserEventDao userEventDao;

  @Autowired
  private UserDao userDao;

  @Test
  public void testAdd() {
    createSimpleEvent();

    List<UserEvent> events = userEventDao.getRepliesForUser(TEST_USER_ID, true, 50, 0, null);

    assertEquals(1, events.size());
  }

  @Test
  public void testInsertTopicUserNotification() {
    userEventDao.insertTopicNotification(TEST_TOPIC_ID, ImmutableList.of(TEST_USER_ID));
  }

  @Test(expected = DuplicateKeyException.class)
  public void testInsertTopicUserNotificationDup() {
    userEventDao.insertTopicNotification(TEST_TOPIC_ID, ImmutableList.of(TEST_USER_ID));
    userEventDao.insertTopicNotification(TEST_TOPIC_ID, ImmutableList.of(TEST_USER_ID));
  }

  private void createSimpleEvent() {
    userEventDao.addEvent(
            UserEventFilterEnum.TAG.toString(),
            TEST_USER_ID,
            false,
            TEST_TOPIC_ID,
            null,
            null
    );
  }

  @Test
  public void testAddRemove() {
    createSimpleEvent();

    List<UserEvent> events = userEventDao.getRepliesForUser(TEST_USER_ID, true, 50, 0, null);

    assertEquals(1, events.size());

    userEventDao.deleteTopicEvents(ImmutableList.of(TEST_TOPIC_ID));

    List<UserEvent> eventsAfterDelete = userEventDao.getRepliesForUser(TEST_USER_ID, true, 50, 0, null);

    assertEquals(0, eventsAfterDelete.size());
  }

  @Test
  public void testRemoveSyntax() {
    userEventDao.deleteTopicEvents(ImmutableList.of(TEST_TOPIC_ID));
  }

  @Test
  public void testRecalc() {
    createSimpleEvent();
    assertEquals(1, userDao.getUser(TEST_USER_ID).getUnreadEvents());
    List<Integer> affected = userEventDao.deleteTopicEvents(ImmutableList.of(TEST_TOPIC_ID));
    assertEquals(1, affected.size());
    assertEquals(1, userDao.getUser(TEST_USER_ID).getUnreadEvents());
    userEventDao.recalcEventCount(ImmutableList.of(TEST_USER_ID));
    assertEquals(0, userDao.getUser(TEST_USER_ID).getUnreadEvents());
  }
}

File: src/test/scala/ru/org/linux/user/UserEventDaoIntegrationTest.scala
/*
 * Copyright 1998-2026 Linux.org.ru
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */
package ru.org.linux.user

import munit.FunSuite
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.test.context.{ContextConfiguration, ContextHierarchy}
import ru.org.linux.scalikejdbc.SpringDB
import ru.org.linux.test.TransactionalTestSupport
import scalikejdbc.*

import java.sql.BatchUpdateException

object UserEventDaoIntegrationTest:
  private val TestTopicId = 98075
  private val TestUserId = 32670

@ContextHierarchy(
  Array(
    new ContextConfiguration(value = Array("classpath:database.xml")),
    new ContextConfiguration(classes = Array(classOf[UserEventDaoIntegrationTestConfiguration]))
  ))
class UserEventDaoIntegrationTest extends FunSuite with TransactionalTestSupport:
  @Autowired
  var userEventDao: UserEventDao = scala.compiletime.uninitialized

  @Autowired
  var userDao: UserDao = scala.compiletime.uninitialized

  @Autowired
  var springDB: SpringDB = scala.compiletime.uninitialized

  test("add"):
    createSimpleEvent()

    val events = userEventDao.getRepliesForUser(
      UserEventDaoIntegrationTest.TestUserId,
      showPrivate = true,
      50,
      0,
      UserEventFilterEnum.ALL)

    assertEquals(events.size, 1)

  test("insertTopicUserNotification"):
    userEventDao.insertTopicNotification(
      UserEventDaoIntegrationTest.TestTopicId,
      Seq(UserEventDaoIntegrationTest.TestUserId))

  test("insertTopicUserNotificationDup"):
    intercept[BatchUpdateException] {
      userEventDao.insertTopicNotification(
        UserEventDaoIntegrationTest.TestTopicId,
        Seq(UserEventDaoIntegrationTest.TestUserId))
      userEventDao.insertTopicNotification(
        UserEventDaoIntegrationTest.TestTopicId,
        Seq(UserEventDaoIntegrationTest.TestUserId))
    }

  private def createSimpleEvent(): Unit =
    userEventDao.addEvent(
      eventType = UserEventFilterEnum.TAG.getType,
      userId = UserEventDaoIntegrationTest.TestUserId,
      isPrivate = false,
      topicId = Some(UserEventDaoIntegrationTest.TestTopicId),
      commentId = None,
      message = None
    )

  test("addRemove"):
    createSimpleEvent()
    val events = userEventDao.getRepliesForUser(
      UserEventDaoIntegrationTest.TestUserId,
      showPrivate = true,
      50,
      0,
      UserEventFilterEnum.ALL)
    assertEquals(events.size, 1)
    springDB.localTx {
      userEventDao.deleteTopicEvents(Seq(UserEventDaoIntegrationTest.TestTopicId))
    }
    val eventsAfterDelete = userEventDao.getRepliesForUser(
      UserEventDaoIntegrationTest.TestUserId,
      showPrivate = true,
      50,
      0,
      UserEventFilterEnum.ALL)
    assertEquals(eventsAfterDelete.size, 0)

  test("removeSyntax"):
    springDB.localTx {
      userEventDao.deleteTopicEvents(Seq(UserEventDaoIntegrationTest.TestTopicId))
    }

  test("recalc"):
    createSimpleEvent()
    assertEquals(userDao.getUser(UserEventDaoIntegrationTest.TestUserId).unreadEvents, 1)
    val affected = springDB.localTx {
      userEventDao.deleteTopicEvents(Seq(UserEventDaoIntegrationTest.TestTopicId))
    }
    assertEquals(affected.size, 1)
    assertEquals(userDao.getUser(UserEventDaoIntegrationTest.TestUserId).unreadEvents, 1)
    springDB.localTx {
      userEventDao.recalcEventCount(Seq(UserEventDaoIntegrationTest.TestUserId))
    }
    assertEquals(userDao.getUser(UserEventDaoIntegrationTest.TestUserId).unreadEvents, 0)

  test("resetUnreadReactionGroup"):
    val topicId = springDB.run {
      sql"SELECT topic FROM comments WHERE NOT deleted GROUP BY topic HAVING count(*) >= 2 ORDER BY topic LIMIT 1"
        .map(rs => rs.int("topic"))
        .single
        .apply()
        .get
    }
    val commentIds = springDB.run {
      sql"SELECT id FROM comments WHERE topic=$topicId AND NOT deleted ORDER BY id LIMIT 2"
        .map(rs => rs.int("id"))
        .list
        .apply()
    }
    assertEquals(commentIds.size, 2)

    val firstCommentId = commentIds.head
    val secondCommentId = commentIds(1)
    val unreadBefore = userDao.getUser(UserEventDaoIntegrationTest.TestUserId).unreadEvents
    val maxEventIdBefore = springDB.run {
      sql"SELECT coalesce(max(id), 0) as maxid FROM user_events".map(rs => rs.int("maxid")).single.apply().get
    }

    userEventDao.addEvent(
      UserEventFilterEnum.REACTION.getType,
      UserEventDaoIntegrationTest.TestUserId,
      isPrivate = false,
      topicId = Some(topicId),
      commentId = Some(firstCommentId),
      message = None,
      originUser = Some(userDao.findUserId("maxcom"))
    )
    userEventDao.addEvent(
      UserEventFilterEnum.REACTION.getType,
      UserEventDaoIntegrationTest.TestUserId,
      isPrivate = false,
      topicId = Some(topicId),
      commentId = Some(firstCommentId),
      message = None,
      originUser = Some(userDao.findUserId("svu"))
    )
    userEventDao.addEvent(
      UserEventFilterEnum.REACTION.getType,
      UserEventDaoIntegrationTest.TestUserId,
      isPrivate = false,
      topicId = Some(topicId),
      commentId = Some(secondCommentId),
      message = None,
      originUser = Some(userDao.findUserId("edo"))
    )

    val insertedEvents = springDB.run {
      sql"""SELECT id, comment_id, unread FROM user_events WHERE userid=${UserEventDaoIntegrationTest
          .TestUserId} AND id>${maxEventIdBefore} AND type='REACTION' ORDER BY id"""
        .map(rs => (rs.int("id"), rs.int("comment_id"), rs.boolean("unread")))
        .list
        .apply()
    }
    assertEquals(insertedEvents.size, 3)

    val firstEventId = insertedEvents.head._1
    val lastEventId = insertedEvents(1)._1

    springDB.localTx:
      userEventDao.resetUnreadReactionGroup(
        UserEventDaoIntegrationTest.TestUserId,
        firstEventId,
        lastEventId,
        topicId,
        firstCommentId)

    val unreadFlagsAfterReset =
      springDB
        .run {
          sql"""SELECT id, unread FROM user_events WHERE userid=${UserEventDaoIntegrationTest
              .TestUserId} AND id>${maxEventIdBefore} AND type='REACTION' ORDER BY id"""
            .map(rs => rs.int("id") -> rs.boolean("unread"))
            .list
            .apply()
        }
        .toMap

    assertEquals(unreadFlagsAfterReset(firstEventId), false)
    assertEquals(unreadFlagsAfterReset(lastEventId), false)
    assertEquals(unreadFlagsAfterReset(insertedEvents(2)._1), true)
    assertEquals(userDao.getUser(UserEventDaoIntegrationTest.TestUserId).unreadEvents, unreadBefore + 1)
